include/sys/queued_operation.h

```diff
@@ -0,0 +1,278 @@
+	 * be invoked recursively, possibly with another operation.
+	 * This can cause unbounded stack growth, and requires that
+	 * process() be re-entrant.  Generally the process() function
+	 * should itself be async, with finalization done after
+	 * process() returns.
+	 *
+	 * @param mgr the service that supports queued operations.
+	 *
+	 * @param op the operation that should be initiated.  A null
+	 * pointer is passed if there are no pending operations.
+	 */
+	void (*process)(struct queued_operation_manager *mgr,
+			struct queued_operation *op);
+};
+
+/**
+ * @brief State associated with a manager instance.
+ */
+struct queued_operation_manager {
+	/* Links the operation into the operation queue. */
+	sys_slist_t operations;
+
+	/* Pointer to the functions that support the manager. */
+	const struct queued_operation_functions *vtable;
+
+	/* Lock controlling access to other fields. */
+	struct k_spinlock lock;
+
+	/* The operation that is being processed. */
+	struct queued_operation *current;
+
+	/* Information about the internal state of the manager. */
+	u32_t volatile state;
+};
+
+#define QUEUED_OPERATION_MANAGER_INITIALIZER(_vtable) {	\
+		.vtable = _vtable,			\
+}
+
+/**
+ * @brief Submit an operation to be processed when the service is
+ * available.
+ *
+ * The service process function will be invoked during this call if
+ * the service is available.
+ *
+ * @param mgr a generic pointer to the service instance
+ *
+ * @param op a generic pointer to an operation to be performed.  The
+ * notify field in the provided operation must have been initialized
+ * before being submitted, even if the operation description is being
+ * re-used.  This may be done directly with async_notify API or by
+ * wrapping it in a service-specific operation init function.
+ *
+ * @param priority the priority of the operation relative to other
+ * operations.  Numerically lower values are higher priority.  Values
+ * outside the range of a signed 8-bit integer will be rejected,
+ * except for named priorities like QUEUED_OPERATION_PRIORITY_APPEND.
+ *
+ * @retval -ENOTSUP if callback notification is requested and the
+ * service does not provide a callback translation.  This may also be
+ * returned due to service-specific validation.
+ *
+ * @retval -EINVAL if the passed priority is out of the range of
+ * supported priorities.  This may also be returned due to
+ * service-specific validation.
+ *
+ * @return A negative value if the operation was rejected by the
+ * service or due to other configuration errors.  A non-negative value
+ * indicates the operation has been accepted for processing and
+ * completion notification will be provided.
+ */
+int queued_operation_submit(struct queued_operation_manager *mgr,
+			    struct queued_operation *op,
+			    int priority);
+
+/**
+ * @brief Helper to extract the result from a queued operation.
+ *
+ * This forwards to :cpp:func:`async_notify_fetch_result()`.
```

Comment:
This isn't rendering properly. Looks like a missing `@rst ... @endrst`.

Reply:
I've gotten rid of the RST syntax, fixed the typo, and changed to "like".  Please unresolve anything that isn't satisfactory or add a new comment.